dask/delayed.py

```diff
@@ -230,7 +231,14 @@ def tokenize(*args, **kwargs):
 
 
 @curry
-def delayed(obj, name=None, pure=None, nout=None, traverse=True):
+def delayed(
+    obj: Any,
```

Comment:
have you tried ``delayed(obj: Union[T, Callable[..., T]], ...) -> Delayed[T]``? Not 100% sure it will work.

Reply:
I have not yet. I am going to dive into this later today and give everything you have suggested a shot. Thanks!

Reply:
Agreed that using a type variable here would be nice, and make the annotations much more powerful. I expect that'll require making the `Delayed` and `DelayedLeaf` classes subclasses of `typing.Generic[T]` though, and annotating all their methods appropriately.

Unfortunately I think that'll be tricky, since AFAIK there's still not a reasonable way to type-annotate a proxy object as being a wrapper around another type (https://github.com/python/mypy/issues/5523). For example, it would be really nice if `delayed({}).foobar` failed a typecheck (because `foobar` is not an attribute of `dict`), but I doubt that'll be possible. I also don't think we could describe correct operations that return a different type, such as that 
`delayed([0, 1])` is of type `Delayed[List[int]]`, and `delayed([0, 1])[0]` is of type `Delayed[int]`, even though we know that `List[T].__getitem__` returns type `T`—I don't think the Python type system can describe that complex of type algebra.

So since I don't know if there would be much utility in `Delayed` being a generic type (since we can't use that type parameter for anything useful), what you have here might be the most reasonable approach for now.

Reply:
It can be used to annotate Delayed.compute to return T.

It's however annoying that this will never work without python/mypy#5523:
```python
@delayed
def double(x: int) -> int:
    return x * 2

x = delayed(2)
y = double(x)  # mypy error
```

Reply:
While the lack (python/mypy#5523) will make it impossible to properly type the proxying behavior of `Delayed` objects,
implementing `Delayed` as a `Generic[T]` has massive advantages for code that operates on delayed objects.

With generic support it's possible to implement type signatures for functions accepting and returning delayed objects,
which is significant pain-point now for larger codebases making heavy use of dask.

Consider, for example, an extremely contrived pipeline performing distributing parsing and analysis of the form:

```python
@dataclass
class RawData: ...

@dataclass
class RawModel: ...
    
@dataclass
class ModelResult: ...

def parse(filename: str) -> RawData: ...
    """Parse input file, always succeeds ;)"""
def model(data: RawData) -> Optional[RawModel]: ...
    """Run model over data, sometimes model is invalid and returns None"""
def analysis(data: List[RawModel]) -> ModelResult: ...
    """Combine model results for final analysis"""
def filter(vals: List[Optional[T]]) -> List[T]:
    """Filter missing values from vals"""
    return [v for v in vals if v is not None]
```

This is, in my opinion, the sweet spot for dask. 
You've got a reasonable bit of code written in normal python and just need `delayed` to make the cores go brrr.

With generic delayed, this pipeline can be type checked and self-documented quite easily.

```python
@overload
def delayed(f: Callable[[List[T], R]]) ->
    Callable[
        [Union[List[T], List[Delayed[T]], Delayed[List[T]]],
        Delayed[R]
    ]: ...

@overload
def delayed(f: Callable[[T], R]) ->
    Callable[
        [Union[T, Delayed[T]],
        Delayed[R]
    ]: ...

d_parse = dask.delayed(parse)
d_model = dask.delayed(model)
d_analysis = dask.delayed(analysis)
d_filter = dask.delayed(filter)

def run_analysis(files: List[str]) -> Delayed[ModelResult]:
    return d_analysis(d_filter([d_model(d_parse(file)) for f in files]))
```

Without generic `Delayed`, the entire typing system collapses in the pipeline, leading to subtle-ish bugs of the form:

```python

def delayed(f: Callable[[T], R]) -> Callable[[Union[T, Delayed]], Delayed]: ...

d_parse = dask.delayed(parse)
d_model = dask.delayed(model)
d_analysis = dask.delayed(analysis)
d_filter = dask.delayed(filter)

def run_analysis(files: List[str]) -> Delayed:
    return d_analysis([d_filter(d_model(d_parse(file))) for f in files])
```

If there's any possibility you can begin investigating adding `Generic[T]` to `Delayed` now I would _extremely_ appreciate it,
including but not limited to mild bribery of your choice. ☕ ? 🍺 ? 🎢 ?

Reply:
@asford Thank you for responding. I am playing on submitting a new PR this evening, if all continues to go well.